dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/cache/internal/ZipFileReaper.java

```diff
@@ -200,15 +212,33 @@ public void run() {
                             reaper.reaperLock.waitNS(reapDelay, methodName, "active pending close"); // throws InterruptedException
                         }
                     } catch ( InterruptedException e ) {
-                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled() ) {
+                        if ( doDebug ) {
                             Tr.debug(tc, methodName + " Interrupted!");
                         }
                         break;
                     }
 
+                    reapAt = SystemUtils.getNanoTime();
+                    if ( doDebug ) {
+                        Tr.debug(tc, methodName + " Reap [ " + toRelSec(startAt, reapAt) + " (s) ]");
+                    }
+
+                    if ( reapDelay > 0L ) {
+                        long actualDelay = reapAt - lastReapAt;
+                        if ( actualDelay > reapDelay ) {
+                            long overage = actualDelay - reapDelay;
+                            if ( overage > STALL_LIMIT ) {
```

Comment:
Are we not concerned about the first time though, where reapDelay == -1?  This is an indefinite wait.... what if the first time through is taking much longer than it needs to be?  What exactly breaks the wait out of the wait forever?

Reply:
An indefinite wait occurs when there are no pending closes.  The wait is broken when a new pending close appears.  In this case, the reaper thread should awaken immediately, as the sleep was broken by a call to notify.  For other cases, the sleep is broken after the requested time has expired.

Most often, the initial wait will be short, but it could grow to be quite long, for example, if the server quiesces and sees no activity for an expended period of time: The server should eventually process all pending closes, and the reaper thread will go back to an indefinite wait.

I could have the code set a time when the notification for a first pending close is made, but that would introduce new state, and would be more complex than handling the definite wait case.